tensorflow/python/keras/mixed_precision/experimental/autocast_variable.py

```diff
@@ -185,46 +185,60 @@ def constraint(self):
     return self._variable.constraint
 
   def assign(self, value, use_locking=None, name=None, read_value=True):
-    return self._variable.assign(value, use_locking, name, read_value)
+    assign_op = self._variable.assign(value, use_locking, name, read_value)
+    return create_autocast_variable(assign_op) if read_value else assign_op
```

Comment:
This will fail with DistributionStrategy with at least two devices. The issue is that `MirroredVariable.assign` will (incorrectly) return a `Mirrored` value instead of a `MirroredVariable`. Following the logic, I think MirroredVariable eventually returns a `Mirrored` [here](https://github.com/tensorflow/tensorflow/blob/147de48ad973a6a05e8113af815988014652caf2/tensorflow/python/distribute/values.py#L1637). I'm guessing that with one device, it simply will return a normal non-mirrored variable.

To reproduce, on [this line](https://github.com/tensorflow/tensorflow/blob/147de48ad973a6a05e8113af815988014652caf2/tensorflow/python/keras/mixed_precision/experimental/autocast_variable_test.py#L64), change `['cpu:0']` to `['cpu:0', 'gpu:0']`. Since you do not have a GPU, also add the following two lines to the top of the file:

```
from google3.third_party.tensorflow.python.framework import config
config.set_soft_device_placement(True)
```

Unfortunately, soft placement seems to be broken in Eager mode (I'll ask internally to get this resolved), but the graph test is running first so you'll be able to reproduce the issue. If you can get it to pass in graph mode, chances are it will pass in eager.

I can't think of an elegant way of fixing this, unfortunately. I think for now, you should simply return `assign_op` if it is `Mirrored` (or perhaps if `not resource_variable_ops.is_resource_variable(assign_op)`). This means #34332 won't be fixed in the distributed case for now, but that can be addressed later. Let me know if you can think of a better solution. 

In the long term, we can fix this by having DistributionStrategy return a MirroredVariable instead of Mirrored from MirroredVariable.assign(). @guptapriya can you work on that?

Reply:
I see, thanks for the help debugging this.

> I can't think of an elegant way of fixing this, unfortunately.

Probably a dumb question, but is there a reason why we cannot allow `AutoCastVariable` to also wrap `Mirrored` and loosen the `is_resource_variable` check?

Reply:
The issue is `AutoCastVariable` subclasses from `tf.Variable` so it doesn't make sense to wrap `Mirrored`. It also defines many assignment methods.

We could create a base class, say `AutoCastTensor`, that could wrap Mirrored. `AutoCastVariable` would then subclass `AutoCastTensor` and `tf.Variable`. But this introduces complexity that I'd rather avoid. Fixing this by having MirroredVariable.assign return a MirroredVariable is a better solution.

Reply:
Makes sense, thanks for the explanation. I made the changes in 9ba63466c6877dbba8278ee26641e62603dddef9 to only wrap it if `is_resource_variable(assign_op) is True`.

Reply:
@reedwm are you saying var.assign(..) should return a variable? that's not what tf.Variable docs say though: https://www.tensorflow.org/api_docs/python/tf/Variable#assign